### PR TITLE
Exception handling and order by uts in transaction fetch query

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below:

- Order by uts in transaction fetch query.

- Exception handling in services.

- Exception handling for observer.

- Subscription data and fetch query is no-cache.

Fixes #165

### CODE REVIEW

**Comment by abhayks1** [comment]
Available for review 🍏

File: src/services/ProgressService.ts [hunk 1]
```diff
@@ -66,7 +66,13 @@ export default class ProgressService {
         && message.targetStatus === MessageStatus.Declared
         ) {
           Logger.debug(`Performing progress stake and progress mint for message hash ${message.messageHash}`);
-          return Promise.all([this.progressStake(message), this.progressMint(message)]);
+          const progressStakePromise = this.progressStake(message).catch(function (error: any) {
+            Logger.error('progressStakeError', error);
+          });
+          const progressMintPromise = this.progressMint(message).catch(function (error: any) {
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply src/services/ProgressService.ts]: What happens if progress stake happens and progress mint fails due to insufficient balance or other issues like geth was down also vice versa? 

Should this be retried somehow? We should create a ticket for this.

File: src/subscriptions/GraphClient.ts [hunk 2]
```diff
@@ -61,26 +61,33 @@ export default class GraphClient {
...
-            Logger.debug('Updating UTS');
-            await GraphClient.updateLatestUTS(
-              transactions,
-              response.data,
-              contractEntityRepository,
-            );
+            try {
+              Logger.debug(`Received subscription data ${JSON.stringify(response.data)}`);
+              const transactions: Record<string,
+                Record<string, any>[]> = await fetcher.fetch(response.data);
```
Comment (sarvesh-ost) [inline src/subscriptions/GraphClient.ts]:
```suggestion
              Record<string, any>[]> = await fetcher.fetch(response.data);
```

File: src/services/AcceptStakeRequestService.ts [hunk 3]
```diff
@@ -97,7 +97,11 @@ export default class AcceptStakeRequestService extends Observer<StakeRequest> {
   private async acceptStakeRequests(stakeRequests: StakeRequest[]): Promise<void> {
     const stakeRequestPromises = [];
     for (let i = 0; i < stakeRequests.length; i += 1) {
-      stakeRequestPromises.push(this.acceptStakeRequest(stakeRequests[i]));
+      stakeRequestPromises.push(
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/AcceptStakeRequestService.ts]: Done!

File: src/services/ConfirmStakeIntentService.ts [hunk 4]
```diff
@@ -137,13 +139,16 @@ export default class ConfirmStakeIntentService extends Observer<Gateway> {
...
   ): Promise<string> {
-    Logger.debug(`Generation proof for confirm stake intent for gateway ${this.gatewayAddress} anf message hash ${message.messageHash}`);
+    Logger.debug(`Generation proof for confirm stake intent for gateway ${this.gatewayAddress} and message hash ${message.messageHash}`);
     const proofData = await proofGenerator.getOutboxProof(
       this.gatewayAddress,
       [message.messageHash],
       gateway.lastRemoteGatewayProvenBlockHeight!.toString(10),
       MESSAGE_BOX_OFFSET, // fixme #141
     );
+    if (proofData.storageProof[0].value === '0') {
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/ConfirmStakeIntentService.ts]: It's 0 not 0x0.

Retrying doesn't work here. Main issue was in ProveGateway. In multiple retries ProveGateway is able to update the state DB with latest state.  
With anchor process waiting for higher time, ProveGateway retries has gone down.

File: src/services/ProgressService.ts [hunk 5]
```diff
@@ -66,7 +66,13 @@ export default class ProgressService {
         && message.targetStatus === MessageStatus.Declared
         ) {
           Logger.debug(`Performing progress stake and progress mint for message hash ${message.messageHash}`);
-          return Promise.all([this.progressStake(message), this.progressMint(message)]);
+          const progressStakePromise = this.progressStake(message).catch(function (error: any) {
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/ProgressService.ts]: Done!

File: src/services/ProgressService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/ProgressService.ts]: Yes there needs to be generic logic to retry failed transactions. With Transactions repository/table already in discussion we will be able to achieve that.

File: src/subscriptions/GraphClient.ts (hunk 2 above)
Comment (abhayks1) [reply src/subscriptions/GraphClient.ts]: Done!

File: src/services/ConfirmStakeIntentService.ts [hunk 6]
```diff
@@ -118,6 +118,8 @@ export default class ConfirmStakeIntentService extends Observer<Gateway> {
         .then((transactionHash): void => {
           Logger.debug(`Message: ${message.messageHash} confirm transaction hash: ${transactionHash}`);
           transactionHashes[message.messageHash] = transactionHash;
+        }).catch(function (error: any) {
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply src/services/ConfirmStakeIntentService.ts]: Done!

**Comment by abhayks1** [comment]
PR available for review 🥗

File: src/services/ConfirmStakeIntentService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply src/services/ConfirmStakeIntentService.ts]: OK. It should automatically fix if anchor state root is delayed.

**Pull request merged**